agent/taskresource/credentialspec/credentialspec.go

```diff
@@ -0,0 +1,556 @@
+		return
+	}
+
+	// Check if the resource transition has already finished
+	if cs.appliedStatus <= knownStatus {
+		cs.appliedStatus = resourcestatus.ResourceStatus(CredentialSpecStatusNone)
+	}
+}
+
+// SetAppliedStatus sets the applied status of resource and returns whether
+// the resource is already in a transition
+func (cs *CredentialSpecResource) SetAppliedStatus(status resourcestatus.ResourceStatus) bool {
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	if cs.appliedStatus != resourcestatus.ResourceStatus(CredentialSpecStatusNone) {
+		// return false to indicate the set operation failed
+		return false
+	}
+
+	cs.appliedStatus = status
+	return true
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (cs *CredentialSpecResource) GetKnownStatus() resourcestatus.ResourceStatus {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.knownStatusUnsafe
+}
+
+// StatusString returns the string of the cgroup resource status
+func (cs *CredentialSpecResource) StatusString(status resourcestatus.ResourceStatus) string {
+	return CredentialSpecStatus(status).String()
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (cs *CredentialSpecResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	cs.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (cs *CredentialSpecResource) GetCreatedAt() time.Time {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.createdAt
+}
+
+// getRequiredCredentialSpecs returns the requiredCredentialSpecs field of credentialspec task resource
+func (cs *CredentialSpecResource) getRequiredCredentialSpecs() map[string][]*apicontainer.Container {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.requiredCredentialSpecs
+}
+
+// getExecutionCredentialsID returns the execution role's credential ID
+func (cs *CredentialSpecResource) getExecutionCredentialsID() string {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.executionCredentialsID
+}
+
+// GetName safely returns the name of the resource
+func (cs *CredentialSpecResource) GetName() string {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return ResourceName
+}
+
+// Create is used to create all the credential spec resources for a given task
+func (cs *CredentialSpecResource) Create() error {
+	// To fail fast, check execution role first
+	executionCredentials, ok := cs.credentialsManager.GetTaskCredentials(cs.getExecutionCredentialsID())
+	if !ok {
+		// No need to log here. managedTask.applyResourceState already does that
+		err := errors.New("credential spec resource: unable to find execution role credentials")
+		cs.setTerminalReason(err.Error())
+		return err
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	for credSpecStr, _ := range cs.requiredCredentialSpecs {
```

Comment:
How many credential specs can be used per task? Should the api call be done in parallel?